arch: arc: optimize the float support

vonhust opened the pull request:
* enable float support
* implement z_arch_float_disable
* add arc support in fp_sharing test

Note: for ARC, maybe k_float_enable(z_arch_float_enable) is needed, as there is no automatic way to enable float, e.g. through exception.

Signed-off-by: Wayne Ren <wei.ren@synopsys.com>

Fixes #16770

Commit:
arch: arc: optimize the float support

* enable float support
* implement z_arch_float_disable
* add arc support in fp_sharing test

Signed-off-by: Wayne Ren <wei.ren@synopsys.com>

ioannisg reviewed (commented):
A side comment, @vonhust ; you need to send a patch to update Zephyr documentation, if FLOAT is, now, to be properly supported by ARC, as well.

Commit:
arch: arc: implement z_arch_float_enable

for arc, floating point support cannot be enabled
automatically, so k_float_enable is requred.

z_arch_float_enable is for k_float_enable

Signed-off-by: Wayne Ren <wei.ren@synopsys.com>

Commit:
doc: add doc of arc floating point support

add doc of arc floating point support

Signed-off-by: Wayne Ren <wei.ren@synopsys.com>

ruuddw commented on tests/kernel/fp_sharing/src/float_regs_arc_gcc.h:
```diff
@@ -0,0 +1,114 @@
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef _FLOAT_REGS_ARM_GCC_H
```
copy-paste mistake, please change into _ARC_ to avoid confusion.

vonhust replied:
Fixed

ruuddw commented:
It looks to me as a k_float_enable() would be required if threads need dynamic switching of FPU use. No other way to enable again. Who/when is the disable called? Only by application thread code that doesn't need FPU anymore, or are there other (kernel?) callers?

vonhust commented:
> It looks to me as a k_float_enable() would be required if threads need dynamic switching of FPU use. No other way to enable again. Who/when is the disable called? Only by application thread code that doesn't need FPU anymore, or are there other (kernel?) callers?

I think k_float_enable/z_arch_float_enable is requried for arch which doest support automatically floating point enable through exception, e.g. ARCv2.

 I just add the implementation of z_arch_float_enable for future use, e.g. k_float_enable.. 

High level implemntation of k_float_enable will impact arch like x86, ARM, so  i didn't change here.
@ioannisg @andrewboie  are better candidate to do that.

vonhust commented:
> A side comment, @vonhust ; you need to send a patch to update Zephyr documentation, if FLOAT is, now, to be properly supported by ARC, as well.

@ioannisg already update the doc about arc

ruuddw commented on tests/kernel/fp_sharing/src/float_context.h:
```diff
@@ -88,6 +88,24 @@ struct fp_non_volatile_register_set {
+	u32_t dpfp2l;
+	u32_t dpfp1h;
+	u32_t dpfp1l;
+#endif
```
How about r58, r59, and FPU status?  It's just a test, but without the single precision FP registers included and 'trashed' only the DA variant gets tested. 

I didn't get the difference between 'volatile' and 'non_volatile' register set here, btw. Not relevant for ARC?

vonhust replied:
for r58 r59,  as they are shared with hardware divider, see instructions divu,  in this test,  r58 r59 will be polluted by divu instructions in z_tick_get_32.  So I just remove them

For ctrl and status, they are aux regs, and not 32-bit r/w,  don't match the mechanisam in the fp_sharing tests., so remove them

For fp_sharing test, i think dpfpxx regs are enough to prove floating point context are saved and restored during context switch.

ruuddw replied:
I'd prefer the test to be more explicit in changing all possible FPU context, instead of relying on implicit changes due to a divu call (which might be avoided/changed during maintenance updates of this code).

Also, if r58/r59 can be used by non FP code, shouldn't they be always saved/restored instead of being part of the lazy 'FP_SHARING' context?

ioannisg commented on doc/reference/kernel/other/float.rst:
```diff
@@ -100,6 +100,31 @@ If an ARM thread does not require use of the floating point registers any
 more, it can call :cpp:func:`k_float_disable()`. This instructs the kernel
 not to save or restore its FP context during thread context switching.
 
+On the ARCv2 architecture the kernel treats each thread as a non-user,
```
Now that we have 3 architectures, here, perhaps it would be nice for the reader to see separate paragraphs, for ARC, ARM and x86 Shared FP services. So they can easily focus on the ARCH they want. @dbkinder what do you think?

dbkinder replied:
I think adding some subheadings would help here, before each section that talks about a specific architecture:

- before line 74 add:
  ```
  ARM Cortex-M architecture (with the Floating Point Extension)
  -------------------------------------------------------------
  ```
- before line 103 add:
  ```
  ARCv2 architecture
  ------------------
  ```
- and before line 128 add:
  ```
  x86 architecture
  ----------------
  ```

ioannisg commented on tests/kernel/fp_sharing/testcase.yaml:
```diff
@@ -12,3 +12,10 @@ tests:
     tags: kernel
     timeout: 600
+  kernel.fp_sharing.arc:
+    extra_args: PI_NUM_ITERATIONS=500
```
How did you decide on this number of iterations?

vonhust replied:
I choose this value dependes on the execution time of the test,  not too fast, not too slow.

ioannisg reviewed (approved):
Looks OK to me.
I left a couple of comments/questions.

Nice that both ARM and ARC are getting better support for FP Services in the upcoming Zephyr release

dbkinder commented on doc/reference/kernel/other/float.rst:
```diff
@@ -100,6 +100,31 @@ If an ARM thread does not require use of the floating point registers any
 more, it can call :cpp:func:`k_float_disable()`. This instructs the kernel
 not to save or restore its FP context during thread context switching.
 
+On the ARCv2 architecture the kernel treats each thread as a non-user,
```
```suggestion
On the ARCv2 architecture, the kernel treats each thread as a non-user
```

vonhust replied:
Fixed

dbkinder commented on doc/reference/kernel/other/float.rst:
```diff
@@ -100,6 +100,31 @@ If an ARM thread does not require use of the floating point registers any
 not to save or restore its FP context during thread context switching.
 
+On the ARCv2 architecture the kernel treats each thread as a non-user,
+or FPU user. It's required to tag a thread as a non-user or FP user by the
```
```suggestion
or FPU user and the thread must be tagged by one of the
```

vonhust replied:
Fixed

dbkinder commented on doc/reference/kernel/other/float.rst:
```diff
@@ -100,6 +100,31 @@ If an ARM thread does not require use of the floating point registers any
+the thread's stack. Thus, the kernel does not save or restore the FP
+context of threads that are not using the FP registers. An extra 16 bytes
+(single floating point hardware) or 32 bytes (double floating point hardware)
+bytes of stack space is required to load and store floating point registers.
```
```suggestion
of stack space is required to load and store floating point registers.
```

vonhust replied:
Fixed

ruuddw commented on arch/arc/include/swap_macros.h:
```diff
@@ -86,6 +89,9 @@ extern "C" {
 	ld sp, [r2, _thread_offset_to_sp]
 
 #ifdef CONFIG_FP_SHARING
+	ld r13, [r2, ___thread_base_t_user_options_OFFSET]
```
see also comments in the test code. I think r58/r59 need to be part of the context that's unconditionally saved/restored, if present.

abrodkin replied:
We do need to save/restore `ACCL`/`ACCH` registers all the time BUT only when they exist.
And they exist only if `MPY_OPTION > 6` or `FPU_FMA_OPTION`.

So in case of the Linux kernel we rely on user specified availability of `ACCL`/`ACCH` with run-time check if user setting matches real HW. That way we don’t need to check HW properties on every context save/restore, see:
1. https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/tree/arch/arc/Kconfig#n404
2. https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/tree/arch/arc/include/asm/entry-arcv2.h#n48

vonhust replied:
already fixed

ruuddw reviewed (changes requested):
Please double check ABI and ARC GNU conventions on use of r58/r59 - seems these are used for non FP instructions as well and should not be conditionally saved/restored.

Commit:
arch: arc: add handling for accl regs, r25, r30

* when fpu is configured or mpy_option > 6,
accl regs (r58, r59) will be configured,
they are used by fpu and mac, and are caller
-saved scratch regs, so need to be saved before
jumping to interrupt handlers

* r25 and r30 are also caller-saved scratch reg.

Signed-off-by: Wayne Ren <wei.ren@synopsys.com>

Commit:
arch: arc: typo fixes and comments clean up

typo fixes and comments clean up

Signed-off-by: Wayne Ren <wei.ren@synopsys.com>

vonhust commented:
@ruuddw , i add specific handling for accl regs (r58, r59). So if no other problems, this PR can be merged

ruuddw commented on arch/arc/core/isr_wrapper.S:
```diff
@@ -306,6 +306,16 @@ _skip_sys_power_save_idle_exit:
 	push_s r3
 
+/* according to ARCv2 ISA, r25, r30, r58, r59 are caller-saved
+ * scratch registers, possibly used by interrupt handlers
```
Please remove the comments on accl/acch NOT being handled at the top of this file, it's handled now :-)

vonhust replied:
Fixed

ruuddw reviewed (changes requested):
Looks good now, just a small fix to keep code comments consistent (accl/acch is now handled).

dbkinder reviewed (approved):
+1 for doc changes, thanks.
